### PR TITLE
Reorganizing the seed project to separate provided code from developer code

### PR DESCRIPTION
**Pull request by NinjaManatee**
- Moved default components to srcDefault
- Moved default configurations to configs/application/default
- Added gulpfile-extenstions.js and server-extenstions.js for extending the gulpfile and server respectively
- Added tutorial describing the seed project structure and extending it.

I didn't change any of the components, I only moved them to a different folder, so the review isn't as bad as it looks.

### CODE REVIEW

**Comment by thorsent** [comment]
@JMCCROWELL readme.md & server/server-extensions.md need an editing pass.

**Comment by thorsent** [comment]
@NinjaManatee changes make a lot of sense. "srcDefault" isn't a great name though. How about: src-builtin-components?

**Comment by NinjaManatee** [comment]
I'm open to changing the name, but I would like it to be short and sweet.
There are adapters and components, so maybe *src-built-in*?


Andy Gardner
*Director, Sales Engineering*
p: 800.821.8147  m: 540.808.5302
w: chartiq.com  e: andy@chartiq.com

**Comment by NinjaManatee** [comment]
server-extensions.md has been removed now. I had already merged it into the
readme.


Andy Gardner
*Director, Sales Engineering*
p: 800.821.8147  m: 540.808.5302
w: chartiq.com  e: andy@chartiq.com

**Comment by thorsent** [comment]
@NinjaManatee src-built-in works for me.

File: README.md [hunk 1]
```diff
@@ -1 +1,243 @@
...
+- _build/webpack_ - Includes all of the files used by the seed project to build the application.
+    - _webpack.default.files.entries.json_ - This specifies the entry and output files for the files built for a default Finsemble application.
+    - _webpack.files.entries.json_ - This file is where developer-added files should be listed. This file is empty in the base Finsemble seed project to prevent merge conflicts when updating the seed project.
+- _configs/application_ - This folder contains all of the base configuration for the Finsemble application. _component.json_, _config.json_ and _services.json_ are empty and developer-added configuration should go here. The files in this folder are merged together to build the application configuration. This configuration can be changed at run time using [Dynamic Configuration](https://documentation.chartiq.com/finsemble/tutorial-dynamicConfiguration.html).
+- _configs/openfin_ - Contains the [OpenFin application config](https://openfin.co/documentation/application-config/) used to start up the Finsemble application. The default manifest for development is included, and additional configurations can be placed in this folder.
```
Comment (carterbs) [inline README.md]: Might be worth pointing out that this is also what openfin means when they talk about the application manifest.

File: README.md [hunk 2]
```diff
@@ -1 +1,243 @@
...
+- _configs/other/server-environment-startup.json_ - Used to define the development and production server configurations used by the Finsemble application.
+- _server_ - Contains the server that hosts the build _dist_ folder for development purposes and includes hot reload in the development environment.
```
Comment (carterbs) [inline README.md]: build => built

File: README.md [hunk 3]
```diff
@@ -1 +1,243 @@
...
+|           webpack.files.js
+|           webpack.services.js
+|           
++---configs
+|   +---application
+|   |   |   components.json
+|   |   |   config.json
+|   |   |   services.json
+|   |   |   
+|   |   \---default
```
Comment (carterbs) [inline README.md]: ├──

This is the character you wanted to use.

File: README.md [hunk 4]
```diff
@@ -1 +1,243 @@
...
+            console.log("POST: pre");
+            done();
+        });
+    };
+
+    // Extension of copyStaticFiles
+    let copyStaticFiles = taskMethods.copyStaticFiles 
+    taskMethods.copyStaticFiles = () => {
+        console.log("PRE: copyStaticFiles");
+        const result = copyStaticFiles();
```
Comment (carterbs) [inline README.md]: result is kind of a strange variable name here. made me pause to think, which i don't think we want examples to do very much.

File: README.md [hunk 5]
```diff
@@ -1 +1,243 @@
...
+            );
+
+        done();
+    }
+};
+```
+
+## Extending server functionality
+
+Sometimes during development, it is necessary to create new API for your application. The Finsemble seed project already includes a server to host files locally for development, so it makes sense to use this server to design the new API. To allow for the extension of the server provided in the Finsemble seed project without breaking the upgrade path, the server attempts to import the _server-extensions.js_ file from the _server_ directory of the project. If _server-extensions.js_ exists, it's methods are called by the server to provide additional functionality. The _server-extensions.js_ file should return an object that contains the following functions:
```
Comment (carterbs) [inline README.md]: I don't think this sentence is necessary. APIs don't need to belong on servers.

File: README.md [hunk 6]
```diff
@@ -1 +1,243 @@
...
+Example _gulpfile-extensions.js_:
+```js
+module.exports = taskMethods => {
+    "use strict";
+
+    const gulp = require("gulp");
+
+    // Extension of pre method;
+    let pre = taskMethods.pre 
+    taskMethods.pre = done => {
```
Comment (rcsharp5) [inline README.md]: This is really confusing to me. I have no clue what it is doing.  Calling `pre` twice?

File: README.md [hunk 7]
```diff
@@ -1 +1,243 @@
...
+    };
+
+    // Extension of copyStaticFiles
+    let copyStaticFiles = taskMethods.copyStaticFiles 
+    taskMethods.copyStaticFiles = () => {
+        console.log("PRE: copyStaticFiles");
+        const result = copyStaticFiles();
+        console.log("POST: copyStaticFiles");
+        
+        return result;
```
Comment (rcsharp5) [inline README.md]: Why a return? Nothing is mentioned about what should be returned.

**Comment by NinjaManatee** [comment]
I’m thinking something like a REST API where the have to fetch data and they want to mock it up. If you don’t think it makes sense, I’ll take it out.

File: README.md [hunk 8]
```diff
@@ -1 +1,243 @@
...
+    };
+
+    const post = taskMethods.post;
+    taskMethods.post = done => {
```
Comment (rcsharp5) [inline README.md]: I guess I must not know much about es6. What is going on here with done? I'm not sure how that is working. Maybe we should dumb it down a bit.

File: README.md (hunk 1 above)
Comment (NinjaManatee) [reply README.md]: Does this sound good:
> Contains the [OpenFin application config](https://openfin.co/documentation/application-config/) (also known as the application manifest) used ...

File: README.md (hunk 4 above)
Comment (NinjaManatee) [reply README.md]: I'm changing the example a bit and adding documentation to make it clear that these methods follow the gulp task function definition.

File: README.md (hunk 5 above)
Comment (NinjaManatee) [reply README.md]: I'm thinking a web API (like REST) that the application uses to fetch information from the server. If you don't think this makes sense, I'll remove it.

File: README.md (hunk 6 above)
Comment (NinjaManatee) [reply README.md]: It is saving the original definition of the `pre` function to be called later. Of course, `pre` doesn't really do anything, so if you think that is too confusing, I'll remove it. I've rewritten a bit around this to make it clearer, but let me know if it still isn't good. (commit to follow).

File: README.md (hunk 7 above)
Comment (NinjaManatee) [reply README.md]: The function in `taskMethods`, except `pre` and `post` are gulp task style functions. I've added some documentation around that, and will commit shortly.

File: README.md (hunk 1 above)
Comment (carterbs) [reply README.md]: perfect

File: README.md (hunk 8 above)
Comment (carterbs) [reply README.md]: @NinjaManatee I agree with ryan here. the lack of parentheses make it hard to follow. I know you don't _have_ to have parens if you just have one arg, but I think it makes it clearer that .post is actually a function if there are parens around done.

**Comment by JMCCROWELL** [comment]
@BradleyCarter : Removed hot reload?

**Comment by carterbs** [comment]
@JMCCROWELL Yeah, it's definitely going to be removed in the next release. Working right now on writing a script that will upgrade people from the 2.0 to 2.3.

File: package.json [hunk 9]
```diff
@@ -5,9 +5,13 @@
   "main": "index.js",
   "scripts": {
     "dev": "set NODE_ENV=development&&gulp --gulpfile=./gulpfile.js",
+    "dev-fresh": "set NODE_ENV=development&&gulp --gulpfile=./gulpfile.js dev:run-fresh",
     "prod": "set NODE_ENV=production&&gulp prod --gulpfile=./gulpfile.js",
+    "prod:run": "set NODE_ENV=production&&gulp prod:run --gulpfile=./gulpfile.js",
     "build": "set NODE_ENV=development&&gulp build --gulpfile=./gulpfile.js",
     "build:prod": "set NODE_ENV=production&&gulp build --gulpfile=./gulpfile.js",
+    "build:vendor-manifest": "webpack --config=./build/webpack/webpack.vendor.config.js",
+    "rebuild": "set NODE_ENV=development&&gulp build --gulpfile=./gulpfile.js rebuild",
```
Comment (thorsent) [inline package.json]: @carterbs what's the difference between rebuild and dev-fresh? What about clean?
@carterbs what is build:vendor-manifest?
@NinjaManatee what is prod:run?

File: server/server.js [hunk 10]
```diff
@@ -65,22 +64,44 @@
...
+	const moduleDirectory = path.join(__dirname, "..", "Finsemble");
 	const cacheAge = 0;
 	const outputColor = chalk.yellow;
 	const PORT = process.env.PORT || 3375;
 	// #endregion
 
 	console.log(outputColor(`Server serving from ${rootDir} with caching maxAge = ${cacheAge} ms.`));
 
-	const options = { maxAge: cacheAge };
+	let options = { maxAge: cacheAge };
```
Comment (thorsent) [inline server/server.js]: @NinjaManatee @carterbs JPM is using our server in production. Others might as well. We should have a "production" mode which has more appropriate caching (24 hours).

We should also ensure that gzip compression is enabled.

File: migration/2.3/migrationScripts/migrate.js [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        "babel-minify-webpack-plugin": "^0.3.0"
+    }
+    existingPackage.devDependencies = Object.assign(existingPackage.devDependencies, newPackages);
+    const content = JSON.stringify(existingPackage, null, '\t');
+    fs.writeFileSync(existingPackagePath, content, 'utf8');
+    done();
+}
+
+
+function addVendorBundle(cb) {
```
Comment (thorsent) [inline migration/2.3/migrationScripts/migrate.js]: @carterbs let's investigate having the windowHeader load the vendor.bundle.js. That way we won't have to change finsemble-cli immediately.

File: migration/2.3/migrationScripts/migrate.js [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+            messageOnComplete: "Copied old gulpfile.",
+            deleteAfterCopy: true
+        },
+        {
+            oldPath: path.join(__dirname, "../gulpfile.js"),
+            newPath: path.join(PROJECT_ROOT, "gulpfile.js"),
+            messageOnComplete: "Copied new gulpfile.",
+        },
+        {
+            oldPath: path.join(PROJECT_ROOT, "build-original/webpack/webpack.files.entries.json"),
```
Comment (thorsent) [inline migration/2.3/migrationScripts/migrate.js]: @carterbs what about the new webpack.files.entries.default.json? I guess the customer will just have duplicates of those components? What would be the right way for a customer to *remove* a component that they don't want if it's in the default.json? I'm not sure what the right answer is for that one.

+ @NinjaManatee

File: build/webpack/defaultWebpackConfig.js [hunk 13]
```diff
@@ -16,15 +17,36 @@ const plugins =
...
+		}
+	}));
+
+	try {
+		const VENDOR_MANIFEST = require('./vendor-manifest.json');
+		plugins.push(new DllReferencePlugin({
+			manifest: VENDOR_MANIFEST
+		}));
+	} catch (e) {
+		console.error(`[WEBPACK ERROR:] You have not generated a vendor-manifest for your webpack configuration. This is an important optimization that reduces build times by 30-40%. Please run "npm run build:vendor-manifest", and then run "npm run dev" once more. You are required to build the vendor manifest when you delete your dist folder, when your node modules update, or when you update the Finsemble Seed project.`);
```
[the remark this answers is not in the document]
Comment (carterbs) [reply build/webpack/defaultWebpackConfig.js]: I added the error before I made the vendor-manifest to be built before the files are.

File: build/webpack/vendor.js [hunk 14]
```diff
@@ -0,0 +1,10 @@
+//Add to this file any modules that are used across many of your component bundles.
+
+const React = require("react"),
+    ReactDom = require("react-dom"),
+    FinsembleControls = require("@chartiq/finsemble-react-controls"),
```
[the remark this answers is not in the document]
Comment (carterbs) [reply build/webpack/vendor.js]: yeah. But each of the sample components uses the react-controls, so they would break too. If the client didn't have the controls, they could just delete this line.

File: gulpfile.js [hunk 15]
```diff
@@ -108,64 +109,46 @@
...
-				if (webpackServicesConfig) {
-					// Webpack config for services exists. Build it
-					webpack(webpackServicesConfig, done);
-				} else {
-					done();
-				}
+			//Requires are done in the function in this way because webpack.files.js will error out if there's no vendor-manifest. The first webpack function generates the vendor manifest.
+			const webpackVendorConfig = require("./build/webpack/webpack.vendor.config.js")
+			webpack(webpackVendorConfig, (err, stats) => {
+				const webpackFilesConfig = require("./build/webpack/webpack.files.js")
```
[the remark this answers is not in the document]
Comment (carterbs) [reply gulpfile.js]: files includes components and adapters. We _could_ split it. Not sure what the value would be.

[the remark this answers is not in the document]
Comment (carterbs) [reply gulpfile.js]: Will talk with andy

File: gulpfile.js [hunk 16]
```diff
@@ -295,16 +278,19 @@
...
-				"clean",
-				taskMethods.copyStaticFiles,
 				taskMethods.buildWebpack,
 				taskMethods.buildSass,
 				taskMethods.buildAngular));
 
+		/**
+		 * Wipes the babel cache and webpack cache, clears dist, rebuilds the application.
+		 */
+		gulp.task("rebuild", gulp.series("clean", "build"));
```
[the remark this answers is not in the document]
Comment (carterbs) [reply gulpfile.js]: just an inconsistency. will fix.

File: migration/2.3/README.md [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+## Client
+1. Unzip the migration folder provided by ChartIQ into the root of your finsemble project.
+2. From the root of that project, run the following command:
+```javascript
+node migration/2.3/migrationScripts/migrate.js
+```
+3. This may take several minutes, mostly because we will have to copy the node_modules into your project. While it's running, feel free to read the notes below.
+
+## Notes on the new seed organization
```
[the remark this answers is not in the document]
Comment (carterbs) [reply migration/2.3/README.md]: I was hoping you'd dump that document you were working on.

File: gulpfile.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (carterbs) [reply gulpfile.js]: Actually, rebuild is just a fresh rebuil.d dev-fresh does a fresh rebuild and starts the server.

File: migration/2.3/migrationScripts/migrate.js (hunk 11 above)
Comment (carterbs) [reply migration/2.3/migrationScripts/migrate.js]: I'll look, but I'm not optimistic. Webpack adds an immediately invoked function into its files that look for the vendor bundle.

File: gulpfile-extensions.md [hunk 18]
```diff
@@ -1,86 +0,0 @@
...
-- `clean` - Cleans the project directory (calls the `clean` method).
-- `build` - Builds the application in the distribution directory (calls "clean" task and, `copyStaticFiles`, `buildWebpack` and `buildSASS` functions).
-- `prod` - Builds the application and starts the server to host it (calls the "build" task then the `startServer` function).
-- `prod:run` - Builds the application, starts the server and launches the Finsemble application (calls the "prod" task then `launchApplication`).
-- `dev:run` - Builds the application, starts the server, launches the Finsemble application and watches for file changes (calls "prod:run" task then the `watchFiles` function).
-- `default` - Specifies the default task to run if no task is passed in (calls the "dev:run" task).
-
-If you wish to add additional task or function calls to the existing tasks, they can be redefined in the `post` method call.
-
-Below is code that can be used as a template for _gulpfile-extensions.js_:
```
[the remark this answers is not in the document]
Comment (JMCCROWELL) [reply gulpfile-extensions.md]: I agree. I also think it breaks up the flow of discussing how the extensions file works. I'm cutting this.
